File: packages/python/awaithumans/types/task.py
"""Task-related types — status, options, record."""

from __future__ import annotations

import enum
from datetime import datetime
from typing import Any, Union

from pydantic import BaseModel, ConfigDict, Field

from awaithumans.types.routing import AssignTo
from awaithumans.types.verification import VerifierConfig, VerifierResult


class TaskStatus(str, enum.Enum):
    CREATED = "created"
    NOTIFIED = "notified"
    ASSIGNED = "assigned"
    IN_PROGRESS = "in_progress"
    SUBMITTED = "submitted"
    VERIFIED = "verified"
    COMPLETED = "completed"
    REJECTED = "rejected"
    TIMED_OUT = "timed_out"
    CANCELLED = "cancelled"
    VERIFICATION_EXHAUSTED = "verification_exhausted"


TERMINAL_STATUSES = frozenset({
    TaskStatus.COMPLETED,
    TaskStatus.TIMED_OUT,
    TaskStatus.CANCELLED,
    TaskStatus.VERIFICATION_EXHAUSTED,
})


class AwaitHumanOptions(BaseModel):
    """Options for the await_human() call."""

    task: str = Field(description="Human-readable description of the task.")
    payload_schema: type[BaseModel] = Field(description="Pydantic model class for the payload.")
    payload: BaseModel = Field(description="The data sent to the human.")
    response_schema: type[BaseModel] = Field(description="Pydantic model class for the response.")
    timeout_seconds: int = Field(
        ge=60,
        le=2_592_000,
        description="Timeout in seconds. Min: 60 (1 minute). Max: 2,592,000 (30 days).",
    )
    assign_to: AssignTo | None = Field(default=None, description="Who should handle this task.")
    notify: list[str] | None = Field(default=None, description='E.g., ["slack:#ops", "email:a@b.com"]')
    verifier: VerifierConfig | None = Field(default=None, description="AI verification config.")
    idempotency_key: str | None = Field(default=None, description="Explicit idempotency key.")
    redact_payload: bool = Field(default=False, description="If true, audit log hides payload.")

    model_config = ConfigDict(arbitrary_types_allowed=True)


class TaskRecord(BaseModel):
    """A task as returned by the SDK/API."""

    id: str
    idempotency_key: str
    task: str
    payload: Any
    payload_schema: dict
    response_schema: dict
    status: TaskStatus
    assign_to: Any | None = None
    response: Any | None = None
    verifier_result: VerifierResult | None = None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    timed_out_at: datetime | None = None
    timeout_seconds: int
    redact_payload: bool = False

File: packages/typescript-sdk/src/types/task.ts
/**
 * Task-shaped types — what the SDK consumes and produces.
 *
 * Mirrors `awaithumans/types/task.py` on the Python side. Keeping the
 * two in lockstep is what lets the cross-language documentation hold
 * up: `AwaitHumanOptions` reads the same in either tongue.
 */

import type { ZodType } from "zod";
import type { JsonSchema7Type } from "zod-to-json-schema";

import type { AssignTo } from "./routing";
import type { VerifierConfig, VerifierResult } from "./verification";

// ─── Core primitive ────────────────────────────────────────────────────

export interface AwaitHumanOptions<TPayload, TResponse> {
	/** Human-readable description of the task. */
	task: string;

	/** Zod schema for the payload — drives the UI the human sees. */
	payloadSchema: ZodType<TPayload>;

	/** The data sent to the human. Must conform to payloadSchema. */
	payload: TPayload;

	/** Zod schema for the response — drives the response form. */
	responseSchema: ZodType<TResponse>;

	/**
	 * Timeout in milliseconds. REQUIRED — no default.
	 * Min: 60,000 (1 minute). Max: 2,592,000,000 (30 days).
	 */
	timeoutMs: number;

	/** Who should handle this task. Optional — defaults to the server's default pool. */
	assignTo?: AssignTo;

	/** Notification channels. E.g., ["slack:#ops", "email:sara@co.com"]. */
	notify?: string[];

	/** AI verifier config — sent to the server for execution. Optional — no verifier = human answer trusted. */
	verifier?: VerifierConfig;

	/** Explicit idempotency key. Defaults to content hash (direct mode) or engine identity (durable adapters). */
	idempotencyKey?: string;

	/** If true, audit log hides the payload body. */
	redactPayload?: boolean;

	/** Server URL override. Defaults to AWAITHUMANS_URL env var or http://localhost:3001. */
	serverUrl?: string;
}

// ─── Task state machine ────────────────────────────────────────────────

export type TaskStatus =
	| "created"
	| "notified"
	| "assigned"
	| "in_progress"
	| "submitted"
	| "verified"
	| "completed"
	| "rejected"
	| "timed_out"
	| "cancelled"
	| "verification_exhausted";

export const TERMINAL_STATUSES: readonly TaskStatus[] = [
	"completed",
	"timed_out",
	"cancelled",
	"verification_exhausted",
] as const;

// ─── Task record ───────────────────────────────────────────────────────

export interface TaskRecord {
	id: string;
	idempotencyKey: string;
	task: string;
	payload: unknown;
	payloadSchema: JsonSchema7Type;
	responseSchema: JsonSchema7Type;
	status: TaskStatus;
	assignTo?: AssignTo;
	response?: unknown;
	verifierResult?: VerifierResult;
	createdAt: Date;
	updatedAt: Date;
	completedAt?: Date;
	timedOutAt?: Date;
	timeoutMs: number;
	redactPayload: boolean;
}
